### DISCUSSION TITLE
Supabase CLI has a problem with circular dependencies by default

### DISCUSSION THREAD

**Discussion by bennik88** [supabase/supabase#29993] [Questions]
When trying to get my production database data into my local development database using `npx supabase db dump -f supabase/seed.sql --data-only` command I get the following error: 
```bash
Dumping data from remote database...
pg_dump: warning: there are circular foreign-key constraints on this table:
pg_dump: detail: key
pg_dump: hint: You might not be able to restore the dump without using --disable-triggers or temporarily dropping the constraints.
pg_dump: hint: Consider using a full dump instead of a --data-only dump to avoid this problem.
```
Upon researching I found out that `pgsodium`is causing the issue. I don't understand why this it is not mentioned in the docs as supabase comes with pgsoodium by default.
I tried to identify the circular dependency with multiple different SQL queries (below), but only ended up being able to identify pgsodium for causing it.
This is hilarious as the supabse cli seems not to provide a way to dump the data and the schema together but only after each other which makes it useless. So I have to use Postgres native functions anyway and that would cause a lot of problems as this would include the `auth` and `storage` schmea by default, as it seems like the only solution is to dump the schema and data in one go or ignore triggers. Link to the supabase docs [here](https://supabase.com/docs/reference/cli/supabase-db-reset#:~:text=on%20branch%20main.-,supabase%20db%20dump,-Dumps%20contents%20from)

Or am I missing something here?
What would be the best solution to this problem?

**My ways to identify problems with circular foreign key relationships:**
```sql
-- Query to List Foreign Key Relationships
SELECT
  tc.constraint_name, 
  tc.table_name AS child_table,
  kcu.column_name AS child_column,
  ccu.table_name AS parent_table,
  ccu.column_name AS parent_column
FROM 
  information_schema.table_constraints AS tc 
  JOIN information_schema.key_column_usage AS kcu
    ON tc.constraint_name = kcu.constraint_name
  JOIN information_schema.constraint_column_usage AS ccu
    ON ccu.constraint_name = tc.constraint_name
WHERE constraint_type = 'FOREIGN KEY';

-- Automated Cycle Detection
WITH RECURSIVE fk_cycles AS (
  SELECT 
    kcu1.table_name AS child_table,
    ccu1.table_name AS parent_table,
    ARRAY[kcu1.table_name] AS path,
    false AS cycle
  FROM 
    information_schema.table_constraints tc1
    JOIN information_schema.key_column_usage kcu1
      ON tc1.constraint_name = kcu1.constraint_name
    JOIN information_schema.constraint_column_usage ccu1
      ON tc1.constraint_name = ccu1.constraint_name
  WHERE 
    tc1.constraint_type = 'FOREIGN KEY'
  
  UNION ALL

  SELECT 
    kcu2.table_name AS child_table,
    ccu2.table_name AS parent_table,
    path || kcu2.table_name,
    kcu2.table_name = ANY(path) AS cycle
  FROM 
    fk_cycles
    JOIN information_schema.table_constraints tc2
      ON fk_cycles.parent_table = tc2.table_name
    JOIN information_schema.key_column_usage kcu2
      ON tc2.constraint_name = kcu2.constraint_name
    JOIN information_schema.constraint_column_usage ccu2
      ON tc2.constraint_name = ccu2.constraint_name
  WHERE 
    tc2.constraint_type = 'FOREIGN KEY'
    AND NOT cycle
)
SELECT * FROM fk_cycles WHERE cycle;


-- Identify foreign key constraints where a table is referencing itself—that is, self-referencing foreign keys
SELECT conrelid::regclass AS table_name, conname AS constraint_name
FROM pg_constraint
WHERE contype = 'f' AND conrelid = confrelid;
```

**Comment by cam-aeynaud** [comment]
I am having the same issue, I have roles.sql, schema.sql and seed.sql files and I get the same warning.
On top of your warning, and no rows in the local development, I get this error:

`failed to send batch: ERROR: syntax error at end of input (SQLSTATE 42601)`

Even though I imported the seed file from doing `npx supabase db dump --data-only -f seed.sql`
